File: week3/이지훈/순위검색.java
import java.util.*;

public class 순위검색 {

    private static Map<String, List<Integer>> conditionToScores = new HashMap<>();

    public List<Integer> solution(String[] infos, String[] queries) {
        List<Integer> results = new ArrayList<>();

        for (String info : infos) {
            getComb(0, info.split(" "), "");
        }

        for (String condition : conditionToScores.keySet()) {
            Collections.sort(conditionToScores.get(condition));
        }

        for (String query : queries) {
            String separationStr = query.replaceAll(" and ", "");
            String[] arr = separationStr.split(" ");

            String condition = arr[0];
            int score = Integer.parseInt(arr[1]);

            results.add(binarySearch(condition, score));
        }

        return results;
    }

    private static void getComb(int depth, String[] infos, String condition) {
        if (depth == 4) {
            if (!conditionToScores.containsKey(condition)) {
                conditionToScores.put(condition, new ArrayList<>());
            }

            conditionToScores.get(condition).add(Integer.parseInt(infos[4]));
            return;
        }

        getComb(depth + 1, infos, condition + "-");
        getComb(depth + 1, infos, condition + infos[depth]);
    }

    private static int binarySearch(String condition, int score) {
        if(!conditionToScores.containsKey(condition)) return 0;

        List<Integer> scores = conditionToScores.get(condition);
        int start = 0;
        int end = scores.size() - 1;

        while (start <= end) {
            int mid = (start + end) / 2;

            if (scores.get(mid) < score) start = mid + 1;
            else end = mid - 1;
        }

        return scores.size() - start;
    }
}

File: week3/한지희/순위 검색.kt
import java.util.*
import kotlin.collections.ArrayList

// 개발언어: cpp, java, python
// 지원 직군: backend, frontend
// 지원 경력구분: junior, senior
// 소울푸드: chicken, pizza
// 점수

class RankSearch {
    private lateinit var usersHashMap: MutableMap<String, ArrayList<Int>>

    fun solution(info: Array<String>, query: Array<String>): IntArray {
        val answer = IntArray(query.size)
        usersHashMap = mutableMapOf()

        info.forEach { applicant ->
            val stringTokenizer = StringTokenizer(applicant)
            val case = Array(4) { Array(2) { "-" } }
            repeat(4) { case[it][0] = stringTokenizer.nextToken() }
            val score = Integer.parseInt(stringTokenizer.nextToken())
            findAllQueryCase(case, "", score, 0)
        }

        usersHashMap.forEach { it.value.sort() }

        query.forEachIndexed { index, queryLine ->
            val stringTokenizer = StringTokenizer(queryLine)
            var queryExceptAnd = ""

            for (i in 0..6) {
                val condition = stringTokenizer.nextToken()
                if (i % 2 == 0)
                    queryExceptAnd += condition
            }

            if (usersHashMap[queryExceptAnd] == null) answer[index] = 0
            else answer[index] = lowerBound(
                usersHashMap[queryExceptAnd]!!,
                Integer.parseInt(stringTokenizer.nextToken())
            )
        }
        return answer
    }

    private fun findAllQueryCase(case: Array<Array<String>>, key: String, score: Int, depth: Int) {
        if (depth == 4) {
            if (usersHashMap[key] == null) {
                usersHashMap[key] = arrayListOf()
            }
            usersHashMap[key]?.add(score)
            return
        }
        findAllQueryCase(case, key + case[depth][0], score, depth + 1)
        findAllQueryCase(case, key + case[depth][1], score, depth + 1)
    }

    private fun lowerBound(scores: List<Int>, targetScore: Int): Int {
        var low = 0
        var high = scores.size - 1
        while (low <= high) {
            val mid = (low + high) / 2
            if (scores[mid] < targetScore) low = mid + 1
            else high = mid - 1
        }
        return scores.size - low
    }

    fun timeoutSolution(info: Array<String>, query: Array<String>): IntArray {
        val answer = IntArray(query.size)
        query.forEachIndexed { index, queryString ->
            var count = 0
            val queryArray = queryString.split(" ")

            info.forEach { applicant ->
                val applicantArray = applicant.split(" ")
                if (
                    (queryArray[0] == applicantArray[0] || queryArray[0] == "-") && // language
                    (queryArray[2] == applicantArray[1] || queryArray[2] == "-") && // job
                    (queryArray[4] == applicantArray[2] || queryArray[4] == "-") && // career
                    (queryArray[6] == applicantArray[3] || queryArray[6] == "-") && // food
                    (queryArray[7].toInt() <= applicantArray[4].toInt()) // score
                ) count++
            }

            answer[index] = count
        }
        return answer
    }
}